addpkg(main/vulkan-wrapper-android): Android Vulkan wrapper

lfdevs opened the pull request:
Original PR by @xMeM : https://github.com/termux/termux-packages/pull/22500

---

- **Robert Kirkman** fixed some build errors when using the latest Termux build system: https://github.com/termux/termux-packages/pull/22500#issuecomment-3335562426
- **nicman23** fixed physical device enumeration for PowerVR GPUs (especially on the DXT-48 GPU equipped in the Pixel 10 series) (https://github.com/lfdevs/termux-packages/issues/4): https://gitlab.freedesktop.org/nicman23/mesa/-/commit/d3eec9f0c4a45a96dc6ea7cb4f469e49cf64d3ed

xMeM committed:
addpkg(main/vulkan-wrapper-android): 25.0.0

lfdevs commented:
You can find the build artifacts here: https://github.com/lfdevs/termux-packages/actions/runs/19142566199  
Specifically, look for files in the Artifacts section that start with `debs`, such as `debs-aarch64-5db7b1b4a39b132466400aca0275fbb1f4a6d30f`. After downloading and extracting it, you’ll obtain `vulkan-wrapper-android_25.0.0_aarch64.deb`.  

Since GitHub Actions artifacts expire after 90 days, I’m providing the extracted `.deb` package directly here:

[vulkan-wrapper-android_25.0.0.zip](https://github.com/user-attachments/files/23397764/vulkan-wrapper-android_25.0.0.zip)

lfdevs commented:
You can refer to the following commands to enable hardware acceleration using Android's Vulkan driver inside a **Termux native** container:

```bash
# Copy or download `vulkan-wrapper-android_25.0.0_aarch64.deb` into the current directory

# Remove packages that may conflict with this guide
pkg remove mesa-zink* vulkan-loader-android
# Install required packages
pkg update && pkg upgrade
pkg install x11-repo
pkg install termux-x11-nightly mesa
pkg install ./vulkan-wrapper-android_25.0.0_aarch64.deb
# Set environment variables
export DISPLAY=:0 MESA_LOADER_DRIVER_OVERRIDE=zink
# Start Termux:X11
kill -9 $(pgrep -f "termux.x11") 2>/dev/null
termux-x11 :0 -dpi 96 &

# Optional steps below
# Install related utilities
pkg install mesa-demos vulkan-tools vkmark glmark2
# Check GLX info
glxinfo
# Check Vulkan info
vulkaninfo
# Run performance benchmarks
vkmark && glmark2
```

lfdevs commented:
# Benchmarks
## Overview

|      Device      |   Model    |             SoC              |      GPU      | vkmark | glmark2 (Zink) |
| :--------------: | :--------: | :--------------------------: | :-----------: | -----: | -------------: |
|  Redmi K80 Pro   | 24122RKC7C | Qualcomm Snapdragon 8 Elite  |  Adreno 830   |   8078 |           1149 |
|   Pixel 6 Pro    |   GLU0G    |       Google Tensor G1       | Mali-G78 MP20 |   1047 |            441 |
| Xiaomi Pad 6 Pro | 23046RP50C | Qualcomm Snapdragon 8+ Gen 1 |  Adreno 730   |   3241 |    Not working |
## Detailed test results
### Adreno 830
<details>
<summary>vkmark</summary>

```
=======================================================
    vkmark 2025.01
=======================================================
    Vendor ID:      0x5143
    Device ID:      0x44050001
    Device Name:    Adreno (TM) 830
    Driver Version: 2150760496
    Device UUID:    64b1fb02435100000000010005440000
=======================================================
[vertex] device-local=true: FPS: 9153 FrameTime: 0.109 ms
[vertex] device-local=false: FPS: 7845 FrameTime: 0.127 ms
[texture] anisotropy=0: FPS: 7919 FrameTime: 0.126 ms
[texture] anisotropy=16: FPS: 7943 FrameTime: 0.126 ms
[shading] shading=gouraud: FPS: 7866 FrameTime: 0.127 ms
[shading] shading=blinn-phong-inf: FPS: 7867 FrameTime: 0.127 ms
[shading] shading=phong: FPS: 7833 FrameTime: 0.128 ms
[shading] shading=cel: FPS: 7868 FrameTime: 0.127 ms
[effect2d] kernel=edge: FPS: 8497 FrameTime: 0.118 ms
[effect2d] kernel=blur: FPS: 7878 FrameTime: 0.127 ms
[desktop] <default>: FPS: 8800 FrameTime: 0.114 ms
[cube] <default>: FPS: 7955 FrameTime: 0.126 ms
[clear] <default>: FPS: 7592 FrameTime: 0.132 ms
=======================================================
                                   vkmark Score: 8078
=======================================================
```

</details>
<details>
<summary>glmark2 (Zink)</summary>

```
=======================================================
    glmark2 2023.01
=======================================================
    OpenGL Information
    GL_VENDOR:      Mesa
    GL_RENDERER:    zink Vulkan 1.3(Adreno (TM) 830 (QUALCOMM_PROPRIETARY))
    GL_VERSION:     4.6 (Compatibility Profile) Mesa 25.2.5
    Surface Config: buf=32 r=8 g=8 b=8 a=8 depth=24 stencil=0 samples=0
    Surface Size:   800x600 windowed
=======================================================
[build] use-vbo=false: FPS: 1653 FrameTime: 0.605 ms
[build] use-vbo=true: FPS: 1640 FrameTime: 0.610 ms
[texture] texture-filter=nearest: FPS: 1592 FrameTime: 0.628 ms
[texture] texture-filter=linear: FPS: 1707 FrameTime: 0.586 ms
[texture] texture-filter=mipmap: FPS: 1645 FrameTime: 0.608 ms
[shading] shading=gouraud: FPS: 1462 FrameTime: 0.684 ms
[shading] shading=blinn-phong-inf: FPS: 1453 FrameTime: 0.688 ms
[shading] shading=phong: FPS: 1361 FrameTime: 0.735 ms
[shading] shading=cel: FPS: 1358 FrameTime: 0.737 ms
[bump] bump-render=high-poly: FPS: 1165 FrameTime: 0.859 ms
[bump] bump-render=normals: FPS: 1537 FrameTime: 0.651 ms
[bump] bump-render=height: FPS: 1608 FrameTime: 0.622 ms
[effect2d] kernel=0,1,0;1,-4,1;0,1,0;: FPS: 1158 FrameTime: 0.864 ms
[effect2d] kernel=1,1,1,1,1;1,1,1,1,1;1,1,1,1,1;: FPS: 893 FrameTime: 1.121 ms
[pulsar] light=false:quads=5:texture=false: FPS: 1401 FrameTime: 0.714 ms
[desktop] blur-radius=5:effect=blur:passes=1:separable=true:windows=4: FPS: 526 FrameTime: 1.904 ms
[desktop] effect=shadow:windows=4: FPS: 943 FrameTime: 1.061 ms
[buffer] columns=200:interleave=false:update-dispersion=0.9:update-fraction=0.5:update-method=map: FPS: 395 FrameTime: 2.535 ms
[buffer] columns=200:interleave=false:update-dispersion=0.9:update-fraction=0.5:update-method=subdata: FPS: 518 FrameTime: 1.932 ms
[buffer] columns=200:interleave=true:update-dispersion=0.9:update-fraction=0.5:update-method=map: FPS: 469 FrameTime: 2.137 ms
[ideas] speed=duration: FPS: 591 FrameTime: 1.692 ms
[jellyfish] <default>: FPS: 929 FrameTime: 1.077 ms
[terrain] <default>: FPS: 264 FrameTime: 3.796 ms
[shadow] <default>: FPS: 919 FrameTime: 1.089 ms
[refract] <default>: FPS: 411 FrameTime: 2.437 ms
[conditionals] fragment-steps=0:vertex-steps=0: FPS: 1317 FrameTime: 0.760 ms
[conditionals] fragment-steps=5:vertex-steps=0: FPS: 1312 FrameTime: 0.762 ms
[conditionals] fragment-steps=0:vertex-steps=5: FPS: 1318 FrameTime: 0.759 ms
[function] fragment-complexity=low:fragment-steps=5: FPS: 1319 FrameTime: 0.758 ms
[function] fragment-complexity=medium:fragment-steps=5: FPS: 1282 FrameTime: 0.780 ms
[loop] fragment-loop=false:fragment-steps=5:vertex-steps=5: FPS: 1314 FrameTime: 0.761 ms
[loop] fragment-steps=5:fragment-uniform=false:vertex-steps=5: FPS: 1314 FrameTime: 0.761 ms
[loop] fragment-steps=5:fragment-uniform=true:vertex-steps=5: FPS: 1198 FrameTime: 0.835 ms
=======================================================
                                  glmark2 Score: 1149
=======================================================
```

</details>

### Mali-G78 MP20
<details>
<summary>vkmark</summary>

```
=======================================================
    vkmark 2025.01
=======================================================
    Vendor ID:      0x13B5
    Device ID:      0x92020000
    Device Name:    Mali-G78
    Driver Version: 226496512
    Device UUID:    a77a84dbf6f534ddb5b88a99395ff94b
=======================================================
[vertex] device-local=true: FPS: 1203 FrameTime: 0.831 ms
[vertex] device-local=false: FPS: 1190 FrameTime: 0.840 ms
[texture] anisotropy=0: FPS: 1153 FrameTime: 0.867 ms
[texture] anisotropy=16: FPS: 1135 FrameTime: 0.881 ms
[shading] shading=gouraud: FPS: 1008 FrameTime: 0.992 ms
[shading] shading=blinn-phong-inf: FPS: 957 FrameTime: 1.045 ms
[shading] shading=phong: FPS: 980 FrameTime: 1.020 ms
[shading] shading=cel: FPS: 872 FrameTime: 1.147 ms
[effect2d] kernel=edge: FPS: 917 FrameTime: 1.091 ms
[effect2d] kernel=blur: FPS: 810 FrameTime: 1.235 ms
[desktop] <default>: FPS: 947 FrameTime: 1.056 ms
[cube] <default>: FPS: 1180 FrameTime: 0.847 ms
[clear] <default>: FPS: 1267 FrameTime: 0.789 ms
=======================================================
                                   vkmark Score: 1047
=======================================================
```

</details>
<details>
<summary>glmark2 (Zink)</summary>

```
WARNING: Some incorrect rendering might occur because the selected Vulkan device (Mali-G78) doesn't support base Zink requirements: feats.features.fillModeNonSolid feats.features.shaderClipDistance
=======================================================
    glmark2 2023.01
=======================================================
    OpenGL Information
    GL_VENDOR:      Mesa
    GL_RENDERER:    zink Vulkan 1.4(Mali-G78 (ARM_PROPRIETARY))
    GL_VERSION:     3.2 (Compatibility Profile) Mesa 25.2.5
    Surface Config: buf=32 r=8 g=8 b=8 a=8 depth=24 stencil=0 samples=0
    Surface Size:   800x600 windowed
=======================================================
[build] use-vbo=false: FPS: 503 FrameTime: 1.988 ms
[build] use-vbo=true: FPS: 547 FrameTime: 1.830 ms
[texture] texture-filter=nearest: FPS: 564 FrameTime: 1.775 ms
[texture] texture-filter=linear: FPS: 552 FrameTime: 1.814 ms
[texture] texture-filter=mipmap: FPS: 580 FrameTime: 1.726 ms
[shading] shading=gouraud: FPS: 519 FrameTime: 1.928 ms
[shading] shading=blinn-phong-inf: FPS: 544 FrameTime: 1.841 ms
[shading] shading=phong: FPS: 516 FrameTime: 1.940 ms
[shading] shading=cel: FPS: 528 FrameTime: 1.897 ms
[bump] bump-render=high-poly: FPS: 404 FrameTime: 2.476 ms
[bump] bump-render=normals: FPS: 615 FrameTime: 1.627 ms
[bump] bump-render=height: FPS: 598 FrameTime: 1.674 ms
[effect2d] kernel=0,1,0;1,-4,1;0,1,0;: FPS: 531 FrameTime: 1.886 ms
[effect2d] kernel=1,1,1,1,1;1,1,1,1,1;1,1,1,1,1;: FPS: 399 FrameTime: 2.508 ms
[pulsar] light=false:quads=5:texture=false: FPS: 585 FrameTime: 1.712 ms
[desktop] blur-radius=5:effect=blur:passes=1:separable=true:windows=4: FPS: 261 FrameTime: 3.846 ms
[desktop] effect=shadow:windows=4: FPS: 357 FrameTime: 2.805 ms
[buffer] columns=200:interleave=false:update-dispersion=0.9:update-fraction=0.5:update-method=map: FPS: 312 FrameTime: 3.211 ms
[buffer] columns=200:interleave=false:update-dispersion=0.9:update-fraction=0.5:update-method=subdata: FPS: 412 FrameTime: 2.432 ms
[buffer] columns=200:interleave=true:update-dispersion=0.9:update-fraction=0.5:update-method=map: FPS: 297 FrameTime: 3.371 ms
[ideas] speed=duration: FPS: 314 FrameTime: 3.191 ms
[jellyfish] <default>: FPS: 450 FrameTime: 2.224 ms
[terrain] <default>: FPS: 164 FrameTime: 6.101 ms
[shadow] <default>: FPS: 419 FrameTime: 2.390 ms
[refract] <default>: FPS: 267 FrameTime: 3.755 ms
[conditionals] fragment-steps=0:vertex-steps=0: FPS: 375 FrameTime: 2.672 ms
[conditionals] fragment-steps=5:vertex-steps=0: FPS: 494 FrameTime: 2.026 ms
[conditionals] fragment-steps=0:vertex-steps=5: FPS: 500 FrameTime: 2.002 ms
[function] fragment-complexity=low:fragment-steps=5: FPS: 436 FrameTime: 2.297 ms
[function] fragment-complexity=medium:fragment-steps=5: FPS: 393 FrameTime: 2.549 ms
[loop] fragment-loop=false:fragment-steps=5:vertex-steps=5: FPS: 329 FrameTime: 3.045 ms
[loop] fragment-steps=5:fragment-uniform=false:vertex-steps=5: FPS: 351 FrameTime: 2.854 ms
[loop] fragment-steps=5:fragment-uniform=true:vertex-steps=5: FPS: 482 FrameTime: 2.076 ms
=======================================================
                                  glmark2 Score: 441
=======================================================
```

</details>

### Adreno 730
<details>
<summary>vkmark</summary>

```
=======================================================
    vkmark 2025.01
=======================================================
    Vendor ID:      0x5143
    Device ID:      0x7030002
    Device Name:    Adreno (TM) 730
    Driver Version: 2150002783
    Device UUID:    39e31f0c435100000000020003070000
=======================================================
[vertex] device-local=true: FPS: 3929 FrameTime: 0.255 ms
[vertex] device-local=false: FPS: 4687 FrameTime: 0.213 ms
[texture] anisotropy=0: FPS: 4769 FrameTime: 0.210 ms
[texture] anisotropy=16: FPS: 2957 FrameTime: 0.338 ms
[shading] shading=gouraud: FPS: 2888 FrameTime: 0.346 ms
[shading] shading=blinn-phong-inf: FPS: 3041 FrameTime: 0.329 ms
[shading] shading=phong: FPS: 2950 FrameTime: 0.339 ms
[shading] shading=cel: FPS: 2814 FrameTime: 0.355 ms
[effect2d] kernel=edge: FPS: 2888 FrameTime: 0.346 ms
[effect2d] kernel=blur: FPS: 2844 FrameTime: 0.352 ms
[desktop] <default>: FPS: 2909 FrameTime: 0.344 ms
[cube] <default>: FPS: 2669 FrameTime: 0.375 ms
[clear] <default>: FPS: 2795 FrameTime: 0.358 ms
=======================================================
                                   vkmark Score: 3241
=======================================================
```

</details>
<details>
<summary>glmark2 (Zink)</summary>

```
glx: failed to create drisw screen
failed to load driver: zink
glx: failed to create drisw screen
failed to load driver: zink
glx: failed to create drisw screen
failed to load driver: zink
Error: glXChooseFBConfig() failed
Error: Error: Couldn't get GL visual config!
Error: main: Could not initialize canvas
```

</details>

nightshadecf commented:
@lfdevs Thank you very much for this build, my termux run straight away without problem.
I get the same benchmark score as you did ( on adreno 830, Oneplus Pad 2 Pro ), everything run much smoother now, especially webgl, this is great :smiley:

lfdevs commented:
> with termux having 25.3.3 mesa i am getting segfaults on everything. could you please bump the package versions?

This driver should work fine even when paired with the latest `mesa` from the Termux repository. What GPU model does your test device have? Please try running `vkcube` with the following commands and see if it works successfully:

```bash
kill -9 $(pgrep -f "termux.x11") 2>/dev/null
termux-x11 :0 -dpi 96 &
DISPLAY=':0' VK_DRIVER_FILES=/data/data/com.termux/files/usr/share/vulkan/icd.d/wrapper_icd.aarch64.json vkcube
```

robertkirkman committed:
fix(main/vulkan-wrapper-android): some build failures

The patch is from @robertkirkman: https://github.com/termux/termux-packages/pull/22500#issuecomment-3335562426

nicman23 committed:
fix(main/vulkan-wrapper-android): physical device enumeration for PowerVR GPUs

This patch is from @nicman23: https://gitlab.freedesktop.org/nicman23/mesa/-/commit/d3eec9f0c4a45a96dc6ea7cb4f469e49cf64d3ed

The previous implementation used a fixed-size static array of 16
VkPhysicalDevice handles, which could cause a segfault on certain
drivers (notably PowerVR DXT-48 on Pixel 10 devices) when the
driver's behavior with the pre-allocated array differed from
expectations.

Replace the static array with dynamic allocation:
i. Query the actual device count first by calling
   EnumeratePhysicalDevices with NULL
ii. Allocate the array based on the returned count
iii. Handle VK_INCOMPLETE by reallocating with a larger size
iv. Add proper error handling and memory cleanup

This follows the Vulkan specification more closely and should be
more robust across different driver implementations.

Fixes: segfault on PowerVR DXT-48 GPU

lfdevs commented:
> Nah man I just fixed a minor nullptr, keep me out of this lol

Thanks for your patch, I've just added it, you can now try it on your Pixel 10 Pro to see if it works properly.

https://github.com/lfdevs/termux-packages/actions/runs/21595714278?pr=1#artifacts

hansm629 commented:
@lfdevs 
Tested on the Galaxy S26 Ultra (Adreno 840), it works perfectly!
In fact, it seems to run more stably than the turnip 26.1.0 devel build!

When running Blender 4.5 with a Vulkan backend, it works flawlessly without any errors!

One minor drawback is that a "wrapper" related message appears in both the zink and Vulkan environments.

Could you please ensure that the message below is not displayed in the next build version?

```
WRAPPER: Entering enumerate_physical_device function
WRAPPER: About to call EnumeratePhysicalDevices
``` 

```
╭─[hsm 󰀲 localhost] [~]
╰─❯ vkcube
Selected WSI platform: xcb
WRAPPER: Entering enumerate_physical_device function
WRAPPER: About to call EnumeratePhysicalDevices
Selected GPU 0: Adreno (TM) 840, type: IntegratedGpu
``` 

```
╭─[hsm 󰀲 localhost] [~]
╰─❯ vkmark
WRAPPER: Entering enumerate_physical_device function
WRAPPER: About to call EnumeratePhysicalDevices
WRAPPER: Entering enumerate_physical_device function
WRAPPER: About to call EnumeratePhysicalDevices
=======================================================
    vkmark 2025.01
=======================================================
    Vendor ID:      0x5143
    Device ID:      0x44050A31
    Device Name:    Adreno (TM) 840
    Driver Version: 2150932499
    Device UUID:    47ffac0b435100000000310a05440000
=======================================================
[vertex] device-local=true: FPS: 2453 FrameTime: 0.408 ms
``` 
```
╭─[hsm 󰀲 localhost] [~]
╰─❯ glmark2
ATTENTION: default value of option vblank_mode overridden by environment.
WRAPPER: Entering enumerate_physical_device function
WRAPPER: About to call EnumeratePhysicalDevices
=======================================================
    glmark2 2023.01
=======================================================
    OpenGL Information
    GL_VENDOR:      Mesa
    GL_RENDERER:    zink Vulkan 1.4(Adreno (TM) 840 (QUALCOMM_PROPRIETARY))
    GL_VERSION:     4.6 (Compatibility Profile) Mesa 26.1.0-devel (git-d7814bcad0)
    Surface Config: buf=32 r=8 g=8 b=8 a=8 depth=24 stencil=0 samples=0
    Surface Size:   800x600 windowed
=======================================================
[build] use-vbo=false: FPS: 1137 FrameTime: 0.880 ms
[build] use-vbo=true: FPS: 1124 FrameTime: 0.890 ms
```

hansm629 commented:
@lfdevs 
I tested `Blender5` with a **Vulkan backend** on the **Galaxy S26 Ultra (Adreno 840 GPU)** using `vulkan-wrapper-android`.

Surprisingly, the results were that it worked without any issues in almost all environments.

The Vulkan backend even operated much more stably than `turnip` from the `26.1.0-devel` upstream.

<img width="853" height="610" alt="3" src="https://github.com/user-attachments/assets/a23c31dc-7879-4073-b4c4-c081c45eb028" />

<img width="1920" height="1080" alt="스크린샷_2026-03-19_22-26-35" src="https://github.com/user-attachments/assets/99da728c-e935-439e-a234-bf0b29d33a3d" />
<img width="1920" height="1080" alt="스크린샷_2026-03-19_22-26-52" src="https://github.com/user-attachments/assets/e9ad8f77-a953-439a-8190-d62e8de70f54" />
<img width="1920" height="1080" alt="스크린샷_2026-03-19_22-27-24" src="https://github.com/user-attachments/assets/2242bc68-3a02-46a3-b618-9c6ba785ff62" />
<img width="1920" height="1080" alt="스크린샷_2026-03-19_22-32-42" src="https://github.com/user-attachments/assets/b787c589-e4c8-4d0a-aa65-4b5d6751ed5a" />
<img width="1920" height="1080" alt="스크린샷_2026-03-19_22-43-47" src="https://github.com/user-attachments/assets/fdacb1c6-c04e-495b-84f5-74954dd73f5a" />
<img width="1920" height="1080" alt="스크린샷_2026-03-19_22-45-13" src="https://github.com/user-attachments/assets/03e05e8d-47c4-4dee-b6c2-4a753f31f7e6" />


**- No artifacts or glitches whatsoever in the Blender5 GUI**

**- No issues in general 3D work environments**

**- Rendering works well**

However, artifacts still occur in the 3D viewer when modeling with specific shaders. Compared to the early days of `Blender4.5`, the **Vulkan backend** is truly stable.
<img width="1920" height="1080" alt="스크린샷_2026-03-19_23-26-55" src="https://github.com/user-attachments/assets/ff36d4fc-a994-4930-935d-7af97391f94f" />
<img width="1920" height="1080" alt="스크린샷_2026-03-20_02-20-52" src="https://github.com/user-attachments/assets/dd2d8424-9367-4495-842d-4e65b068e1d6" />


Even if artifacts occur in the 3D viewer, rendering works perfectly without any issues!

However, `Blender5`'s **Vulkan backend** operates stably only in the `Adreno 840 GPU` + `vulkan-wrapper-android `environment.

In the case of the **Galaxy S25 Ultra**'s `Adreno 830 GPU`, perhaps due to the low Vulkan driver version, running `Blender5` with the **Vulkan backend** results in various symptoms, such as GUI artifacts and glitches, as well as various graphic glitches in the 3D viewer.

The rendering results also come out strangely.

From what I've seen, the most stable way to run `Blender` on **Termux Bionic Native** is the combination of

**`Adreno 840 GPU` + `vulkan-wrapper-android` + `Blender5 Vulkan backend`**




@robertkirkman 
I think it will be perfect after just two more updates!

I look forward to the next build of Blender5!
Thank you!

robertkirkman commented:
The newest version of blender is currently 5.0.1,

when new versions of blender appear here, i download and compile them in termux:

https://download.blender.org/release/Blender5.0/

robertkirkman commented:
oh never mind, there is new blender 5.1 actually, here

https://download.blender.org/release/Blender5.1/

I will try to compile blender 5.1

robertkirkman commented:
I tried to compile blender 5.1 but I have this error, I don't know how to fix this but I can try

```
/data/data/com.termux/files/home/.termux-build/blender5/src/intern/libmv/libmv/tracking/track_region.cc:943:17: error: reference to non-static member function must be called; did you mean to call it with no arguments?
  943 |     Vec4 a = Q1.jacobiSvd<Eigen::ComputeThinU | Eigen::ComputeThinV>().solve(Q2);
      |              ~~~^~~~~~~~~
      |                          ()
/data/data/com.termux/files/home/.termux-build/blender5/src/intern/libmv/libmv/tracking/track_region.cc:943:26: error: invalid operands to binary expression ('JacobiSVD<PlainObject>' (aka 'JacobiSVD<Eigen::Matrix<double, -1, -1, 0, -1, -1>>') and 'Eigen::DecompositionOptions')
  943 |     Vec4 a = Q1.jacobiSvd<Eigen::ComputeThinU | Eigen::ComputeThinV>().solve(Q2);
      |              ~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~
/data/data/com.termux/files/home/.termux-build/blender5/src/intern/libmv/libmv/tracking/track_region.cc:943:17: error: 'jacobiSvd' does not name a template but is followed by template arguments
  943 |     Vec4 a = Q1.jacobiSvd<Eigen::ComputeThinU | Eigen::ComputeThinV>().solve(Q2);
      |                 ^        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/data/data/com.termux/files/usr/include/eigen3/Eigen/src/Core/MatrixBase.h:371:35: note: non-template declaration found by name lookup
  371 |     inline JacobiSVD<PlainObject> jacobiSvd(unsigned int computationOptions = 0) const;
      |                                   ^
1 warning and 3 errors generated.
```

robertkirkman commented:
I compiled Blender 5.1, but unfortunately, it has an error and I don't understand how to fix it :crying_cat_face: 

- https://github.com/termux-user-repository/tur/issues/2363

lfdevs commented:
I'm not sure if the test situation under PRoot is useful for Termux Native, but perhaps it can be used for reference:

When I previously tried to build Blender **v5.0.1** (including its dependencies) using a `rockylinux:8` container (Arm64), I had to add [these four patches](https://github.com/lfdevs/blender-linux-arm64/commit/e25fd6a6ad7d1c1e155645b70b663db6d3f97446) for the build to succeed. However, when running the build artifacts on Adreno 830, neither the OpenGL backend nor the Vulkan backend could open Blender's GUI interface (the terminal prompted a segmentation fault).

But when I tried to build Blender **v5.1.0** yesterday, I only added [one patch to exclude ROCm](https://github.com/lfdevs/blender-linux-arm64/commit/963863d57431e01a8365558d70c4257dd412aa19), and it built successfully directly. Moreover, Adreno 830 could run it successfully using both the OpenGL backend (with Freedreno) and the Vulkan backend (with Turnip). Although the OpenGL backend still flickers severely and the Vulkan backend even only displays the window frame, this is also good news for me.

I don't know if this is because Blender upstream [updated the dependencies for 5.1](https://projects.blender.org/blender/blender/pulls/149852) (the other three patches for dependencies are no longer needed), or simply because I changed the source from a GitHub mirror to the official Blender repository.
